Add a property to enable stricter CC entry integrity checks (#32434)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/optimizing-builds/configuration_cache.adoc

Check the method documentation for more details.

[[config_cache:integrity_check]]

> === Using integrity checks to debug store and load issues

Reviewer: 💅 Add empty line after the header

By default, the configuration cache doesn't perform many integrity checks when reading the entry.
This improves performance and reduces entry size.
However, it makes it harder to pinpoint some store and load issues.
For example, the incorrectly stored object may be discovered later, when reading data of some unrelated task.

> It may cause a type mismatch when restoring a value of the property or break some other invariant, resulting in a potentially misleading exception.

Reviewer: 💭 Not a native speaker, but here an opinionated suggestion on how to rephrase the paragraph. Feel free to ignore, pick parts, etc.

```text
Configuration cache avoids extra performance and on-disk space costs by performing minimal integrity checks as it writes and reads the data.
However, it can make it harder to troubleshoot cases when concurrency problems or misbehaving serializers interfere with the data.
For example, an incorrectly written object may be discovered only later when reading data of some unrelated task.
It may cause a type mismatch when restoring a value of the property or break some other invariant, resulting in a potentially misleading error messages. 
```

In order to make debugging easier, it is possible to enable stricter checks that may help to discover discrepancies earlier.
To do that, add the following line to your `gradle.properties`:
[source,properties]
.gradle.properties
----
org.gradle.configuration-cache.integrity-check=true
----

For example, let's look at a type that implements a custom serialization protocol incorrectly:
[source,java]
----
include::{snippetsPath}/configurationCache/integrityChecks/common/buildSrc/src/main/java/User.java[tags=broken-type]
----
<1> `writeObject` serializes both fields.
<2> `readObject` reads only the first field, leaving the lingering data.

Such a type will cause problems when used as part of a task state.
[source,java]
----
include::{snippetsPath}/configurationCache/integrityChecks/common/buildSrc/src/main/java/GreetTask.java[tags=task-with-broken-type]
----

> When running without the integrity check, you may get a cryptic failure message, perhaps accompanied by some induced configuration cache problems:

Reviewer: 🤔 Now that we have a flag, it makes me wonder if we should suggest using it in such cases. Even if it's a heuristic for allow-listed set of exceptions, it would already point users in the right direction of the integrity checks. Otherwise, they are expected to know or read through the manual, and we all know what this usually means.

Author: > Now that we have a flag, it makes me wonder if we should suggest using it in such cases.

We can do it as part of adopting the Problems API in CC.

----
❯ gradle --configuration-cache greet
...
* What went wrong:
Index 4 out of bounds for length 3
----

It might be really hard to connect the invalid index to a serialization issue. Enabling the integrity check gives a much more precise diagnostics:
----
❯ gradle --configuration-cache -Dorg.gradle.configuration-cache.integrity-check=true greet
...
include::{snippetsPath}/configurationCache/integrityChecks/tests/fail.out[]
----
You can immediately see the name of the offending task and even the field that contains the broken data.

> Keep in mind that this attribution is best-effort: it should be accurate in most cases, but rarely can be confused by some byte patterns.

Reviewer: 💭 I think we should add a bold warning that this flag does not work with the data already written to disk. Aka users won't be able, upon failure, check what is wrong with the current entry. Instead, a new entry will be created, due to the flag being part of the key. So any concurrency issues would have to be reproduced again.

Check the method documentation for more details.

[[config_cache:integrity_check]]
=== Using integrity checks to debug store and load issues

> In order to improve performance and reduce entry size, the configuration cache performs a minimal amount of integrity checks as it writes or reads the data.

Reviewer 2: ```suggestion
To reduce entry size and improve performance, the cache performs minimal integrity checks when writing and reading data.
```

> However, it makes it harder to troubleshoot issues caused by concurrency problems or misbehaving serialization logic.

Reviewer 2: ```suggestion
However, this approach can make troubleshooting issues more difficult, especially when dealing with concurrency problems or serialization errors. 
```

> The incorrectly stored object may not be discovered immediately, but it can affect reading data of some other, unrelated task.

Reviewer 2: ```suggestion
An incorrectly stored object may not be detected immediately but could lead to misleading or misattributed errors when reading cached data later.
```

> For example, reading a value of the property may end up with a type mismatch or break some other invariant, resulting in a potentially misleading or misattributed error message.

Reviewer 2: ```suggestion
For example, reading the value of a cached property might cause a type mismatch or break an invariant, resulting in a potentially misleading or misattributed error message.
```

> In order to make debugging easier, it is possible to enable stricter checks that may help to discover discrepancies earlier,

Reviewer 2: ```suggestion
To make debugging easier, Gradle provides an option to enable stricter integrity checks.
```

> at a cost of slowing down cache operations and significantly increasing the size of the entry.

Reviewer 2: ```suggestion
This setting helps identify inconsistencies earlier but may slow down cache operations and increase the cache entry size.
```

> To do that, add the following line to your `gradle.properties`:

Reviewer 2: ```suggestion
To enable stricter integrity checks, add the following line to your `gradle.properties` file:

```

[source,properties]
.gradle.properties
----
org.gradle.configuration-cache.integrity-check=true
----

> For example, let's look at a type that implements a custom serialization protocol incorrectly:

Reviewer 2: ```suggestion
For example, let's look at a type that implements a custom serialization protocol incorrectly:

```

[source,java]
----
include::{snippetsPath}/configurationCache/integrityChecks/common/buildSrc/src/main/java/User.java[tags=broken-type]
----
<1> `writeObject` serializes both fields.
<2> `readObject` reads only the first field, leaving the remaining data in the stream.

> Such a type will cause problems when used as part of a task state, because the configuration cache will try to interpret the remaining unread part of the object as some new value.

Reviewer 2: ```suggestion
Such a type will cause problems when used as part of a task state, because the configuration cache will try to interpret the remaining unread part of the object as some new value:

```

[source,java]
----
include::{snippetsPath}/configurationCache/integrityChecks/common/buildSrc/src/main/java/GreetTask.java[tags=task-with-broken-type]
----

> When running without the integrity check, you may get a cryptic failure message, perhaps accompanied by some induced configuration cache problems:

Reviewer 2: ```suggestion
When running without the integrity check, you may encounter cryptic failure messages, possibly accompanied by induced configuration cache problems:

[source,bash]
```

----
❯ gradle --configuration-cache greet
...
* What went wrong:
Index 4 out of bounds for length 3
----

> It might be really hard to connect the invalid index to a serialization issue. Rerunning the build with the integrity check enabled gives a much more precise diagnostic:

Reviewer 2: ```suggestion
These errors might not immediately point to the root cause, making debugging more challenging. 
It might be really hard to connect an invalid index error to a serialization issue, for example.

Rerunning the build with the integrity check enabled provides a much more precise diagnostic, helping you pinpoint the source of the issue faster:

[source,bash]
```

----
❯ gradle --configuration-cache -Dorg.gradle.configuration-cache.integrity-check=true greet
...
include::{snippetsPath}/configurationCache/integrityChecks/tests/fail.out[]
----

> You can immediately see the name of the offending task and even the field that contains the broken data.

Reviewer 2: ```suggestion

You can immediately see the name of the offending task and the field that contains the broken data.

```

> Keep in mind that this attribution is best-effort: it should be accurate in most cases, but rarely can be confused by some byte patterns.

Reviewer 2: ```suggestion
Keep in mind that this attribution is best-effort: it should be accurate in most cases, but in rare instances, it may be confused by certain byte patterns.
```

[WARNING]
====

> The integrity check relies on extra data stored in the cache, and thus cannot be used to debug an existing corrupted entry written without the check enabled.

Reviewer 2: ```suggestion
The integrity check relies on additional metadata stored within the cache. Therefore, it cannot be used to diagnose entries already corrupted prior to enabling the integrity check.
```

====

> Note, that the current set of integrity checks primarily aims at detecting issues with serialization protocols more than data corruption.

Reviewer 2: ```suggestion
The current integrity checks primarily focus on identifying serialization protocol issues rather than general data corruption. 
```

> It may not be as effective against issues caused by hardware failures like bit rotting or bad sectors.

Reviewer 2: ```suggestion
Consequently, they are less effective against hardware-related problems, such as bit rot or damaged storage sectors.
```

> Because of that and of the extra costs the integrity check brings, we don't recommend to have it always enabled, but consider it a troubleshooting tool.

Reviewer 2: ```suggestion
Due to these limitations and the performance overhead introduced by integrity checks, we recommend enabling them selectively as a troubleshooting measure rather than leaving them permanently enabled.
```
